Make forward declarations with `Ref` safer

AThousandShips opened the pull request:
`Ref`s to forward declared types are context dependent if the destructor is not explicit and in a source file

As seen in: #80278

~~This doesn't affect the current codebase, but makes it unreliable~~

~~g++ seems to define destructors for classes with no explicit destructor when they first are called, meaning that if a class is undefined in that context it will cause errors, and depends entirely on the order of compilation, this makes destructors explicit for classes that have `Ref`s to forward declared classes, I believe I have caught every case (not including cases where the definition occurs in the same header)~~

~~Some of these cases can probably be done by removing the forward declaration instead, also attempted alternative solutions but unfortunately they were unsuccessful~~

<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

lawnjelly commented:
> Refs to forward declared types are unreliable with g++, this ensures destructors are explicit to prevent unpredictable behavior

For the benefit of those just reading this PR, in what sense is "unreliable", and "unpredictable" meant? It builds successfully on some CI platforms and not others? :thinking: 

Due to pushing on #80278 we can unfortunately no longer see the CI run that failed. That's quite difficult for reviewers here in terms of seeing exactly how it is failing.

Am I correct in thinking the problem is that `Ref<Foo>` should `#include "foo.h"`, rather than forward declare `class Foo;`? Because the destructor is called in `Ref` via `memdelete`?

If so, it seems ideally we would like to either:
* To trigger the need for the include file in compilers that don't require it
* Remove the need for the include file in compilers that require it

Is there a danger that removing the destructors as in this PR would leave the problem to crop up again every time someone adds a new destructor?

_(From the sound of it, this wasn't the problem,  but just to check - it wasn't a SCU build that was passing was it? That always strictly speaking needs a regular build before submitting a PR.)_

AThousandShips commented:
It was failing on all the builds with g++ it seems

The unpredictable/unreliable here is that while it currently doesn't cause any problems, the following senario would cause problems:
* Some class X has a `Ref` to a forward declared Y
* Some other class Z references X, but does the correct includes for Y
* Someone makes a new PR, with some class W, which does not include Y, and is compiled *before* Z

The unreliable and unpredictable part, and what makes diagnosing and fixing difficult (as evidenced by the referenced PR) is that this can occur anywhere along the chain, so if for example you make X reference some forward declared Y, and you correctly include Y in the source file for X, *but* some other file uses X you will now get massively cryptic error messages when compiling some seemingly unrelated class

In this case it was:
* `CollisionShape3DGizmoPlugin` has a member `Ref<Gizmo3DHelper>`
* `Node3DEditorPlugin` references `CollisionShape3DGizmoPlugin`
* The error shows up:
```
In file included from editor/plugins/node_3d_editor_plugin.cpp:31:
In file included from editor/plugins/node_3d_editor_plugin.h:34:
In file included from ./editor/editor_plugin.h:34:
In file included from ./core/io/config_file.h:34:
In file included from ./core/io/file_access.h:36:
Error: ./core/object/ref_counted.h:209:29: error: member access into incomplete type 'Gizmo3DHelper'
                if (reference && reference->unreference()) {
                                          ^
./core/object/ref_counted.h:222:3: note: in instantiation of member function 'Ref<Gizmo3DHelper>::unref' requested here
                unref();
                ^
./editor/plugins/gizmos/gpu_particles_collision_3d_gizmo_plugin.h:38:7: note: in instantiation of member function 'Ref<Gizmo3DHelper>::~Ref' requested here
class GPUParticlesCollision3DGizmoPlugin : public EditorNode3DGizmoPlugin {
      ^
./core/object/ref_counted.h:210:4: note: in instantiation of function template specialization 'memdelete<GPUParticlesCollision3DGizmoPlugin>' requested here
                        memdelete(reference);
                        ^
./core/object/ref_counted.h:222:3: note: in instantiation of member function 'Ref<GPUParticlesCollision3DGizmoPlugin>::unref' requested here
                unref();
                ^
editor/plugins/node_3d_editor_plugin.cpp:7890:19: note: in instantiation of member function 'Ref<GPUParticlesCollision3DGizmoPlugin>::~Ref' requested here
        add_gizmo_plugin(Ref<GPUParticlesCollision3DGizmoPlugin>(memnew(GPUParticlesCollision3DGizmoPlugin)));
                         ^
./editor/plugins/gizmos/collision_shape_3d_gizmo_plugin.h:36:7: note: forward declaration of 'Gizmo3DHelper'
class Gizmo3DHelper;
      ^
```

AThousandShips commented:
This is mostly preemptive, as the things *currently* work, but the concern is that seemingly unrelated changes can cause this

lawnjelly commented:
imo it sounds like it just needs an `#include` where using the template functions that requires access to the definition.

Cryptic template error messages is one of the joys of templates unfortunately. :man_shrugging: 

Making it "easier to use" might be something like adding the `#include` in `X.h` in your example, even if not strictly required by X, but I'm not sure I would advise this. Compilation is all about minimizing these kinds of dependency chains. 

The fact it doesn't require the include on some compilers is what is interesting, it would be nice to see a minimal example.

AThousandShips commented:
Yeah I have no clue why MSVC *doesn't* have a problem here, but that just makes it even more insidious an error

lawnjelly commented:
Which builds did it succeed on?

AThousandShips commented:
Tested locally to induce the behavior and it *does* happen on MSVC as well, it just *doesn't* happen on the CI, probably due to what things it does in that build

Here's the job:
https://github.com/godotengine/godot/actions/runs/5767317890/job/15636812267

KoBeWi commented:
> Due to pushing on https://github.com/godotengine/godot/pull/80278 we can unfortunately no longer see the CI run that failed. That's quite difficult for reviewers here in terms of seeing exactly how it is failing.

If you can test locally, this is the commit that was failing: https://github.com/godotengine/godot/commit/c75e58908316fb0ee843c580177928cdb73968de

lawnjelly commented:
Agreed it looks kind of hard to pin down from the error message. Could be down to the compiler, or could be something in the platform specific code.

But it does seem like there are some compiler differences in how they handle forward declarations. This is old but:
https://cboard.cprogramming.com/cplusplus-programming/113393-does-gcc-hate-forward-declarations.html

> What brewbuck says is exactly the reason why VC++ compiles the code, but it's not standard C++.

> The thing is, VC++ doesn't do semantic analysis of templates at all before they're instantiated. It pretty much tokenizes the template body and then stores the tokens away, to be reparsed when the template is instantiated, at the place (and in the context) of the instantiation. The consequence? A type that is incomplete at the template definition site can be used if it's complete at the instantiation site.

> But that's non-compliant behavior. The standard requires the compiler to do all semantic analysis except for a few things that involve actual template parameters. The call to baz.quuz() in the example above does not involve the template parameters at all, and thus the compiler is actually required to refuse to parse the template body.

> (More correctly, the compiler may delay diagnostics until the instantiation, but it must emit the same diagnostics as it would if it didn't.)

lawnjelly commented:
If we go with the idea that the problem is the microsoft compiler might be too tolerant and non-standard, I'm wondering if there is a flag to disable this behaviour.

Could this help?
https://learn.microsoft.com/en-us/cpp/build/reference/permissive-standards-conformance?view=msvc-170

AThousandShips commented:
Added a new commit changing relevant cases to include directly, unsure about some cases so left them as is, as well as cases that were circular

Can decide on a case-by-case basis

There are potential pitfall cases like `MeshInstance3D` which has an empty destructor which could get removed and risk this issue, but left all cases where a destructor already exits in source, where there's no (current) risk

AThousandShips commented:
I know well, but I think there needs to be a balance there, where you shouldn't depend on highly unexpected situations

YuriSizov commented:
I think keeping headers lean is an important goal. Moving things back into them is going to cost in compile times, as well as create more situations for implicit dependencies. What's even worse is that the reason for the include to be there is not documented in any way. Whereas with an empty destructor you can add a comment why it's needed.

Finally, we need to be able to catch these cases on CI, otherwise any effort would be rather pointless if not counter-productive. There is a benefit to the status quo as well, as it did produce an error for you to notice.

AThousandShips commented:
Unsure how we could catch these in CI except when it actually does break, the effort is to ensure that the ability of these classes to compile doesn't depend on some hard to diagnose order of compilation down the line

YuriSizov commented:
Do explicit empty destructors solve the issue for every case?

AThousandShips commented:
Removed the include version (saved if we want part of it) and added a descriptive comment to all the destructors:
```
// Do not remove, kept to prevent forward declaration issues, see:
// https://github.com/godotengine/godot/pull/80330
```

AThousandShips committed:
Prevent unreliable behavior for forward declarations

`Ref`s to forward declared types are unreliable with g++, this ensures
destructors are explicit to prevent unpredictable behavior

akien-mga commented on core/io/pck_packer.cpp:
```diff
@@ -259,3 +259,8 @@ Error PCKPacker::flush(bool p_verbose) {
 	return OK;
 }
+
+PCKPacker::~PCKPacker() {
+	// Do not remove, kept to prevent forward declaration issues, see:
+	// https://github.com/godotengine/godot/pull/80330
+}
```
For the record, I haven't reviewed the discussion in depth, but I'm not too fond of:
- Having to remember to add empty destructors, without a clear criterion to know which classes need it.
- Hardcoding the same comment and link to a GitHub issue everywhere.

AThousandShips replied:
Could enforce destructors universally for safety

AThousandShips replied:
But the criterion is:
Has a `Ref` to a forward declared class

AThousandShips replied:
So to clarify the issue here:
A class, which contains a `Ref` to a forward declared class, must on the first time it is invoked, be it directly or indirectly, have a full declaration available, regardless of the access to that specific class

This differs from pointers

So imagine a class `A` references a class `B` which references a class `C` (which is forward declared), then the scope in which `A` is destructed, even if it does *no* interaction *what so ever* with `C` (all that is done internally in `B`), must contain the full declaration of `C`, if it is the first scope which destructs `C`

YuriSizov replied:
I was looking at some code and realized that we may already have these empty destructors that somehow "fix things", just without any comment or explanation. Throughout the codebase we have empty constructors and destructors present for no apparent reason, probably just as a stylistic choice by some C++ developers. Given what this PR does, I think some of those may accidentally be preventing similar bugs without us even realizing.

So, in my opinion we should definitely ensure this happens in a more organized and obvious manner.

AThousandShips commented:
After doing some digging I've realized that this isn't order dependent at all, it's simply that the destructor chain is called in each local scope and therefore if some class has a forward declaration and contains a member variable that has a `Ref` to that then if any context invokes and creates that destructor down the line it must include that class in the chain, even if it doesn't interact with it at all, and even if the class itself doesn't expose that type

This goes down chains and can be removed several levels, for example if:
* `ClassA` has a forward declared `Ref<Foo>`, and no explicit destructor
* `ClassB` has a `Ref<ClassA>` or `ClassA` as a plain value, and no explicit destructor (even at this level it shouldn't have to care about `Foo` as `ClassA` doesn't expose it in any way and `ClassB` doesn't do anything with `Foo`)
* `ClassC` has a `Ref<ClassB>` or `ClassA` then the scope within which the destructor of *it* is invoked, be it explicit or otherwise, must include `Foo`

Meaning that if, for example, you have a reference to `ClassC` in `main.cpp` then it will cause problems *in* `main.cpp` and the error message will be very cryptic

Will dig through and update this PR and clarify the new findings when I have the time and energy, but turns out this is more specific and critical than I first thought

AThousandShips commented:
Will look into other ways of achieving higher safety with these types in the future
